smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,169 @@
+        allow_overwrite: bool = ...,
+    ) -> t.Callable[[type[LaunchArgBuilder[_T]]], type[LaunchArgBuilder[_T]]]: ...
+    @t.overload
+    def dispatch(
+        self,
+        args: type[LaunchArgBuilder[_T]],
+        *,
+        to_launcher: type[LauncherLike[_T]],
+        allow_overwrite: bool = ...,
+    ) -> None: ...
+    def dispatch(
+        self,
+        args: type[LaunchArgBuilder[_T]] | None = None,
+        *,
+        to_launcher: type[LauncherLike[_T]],
+        allow_overwrite: bool = False,
+    ) -> t.Callable[[type[LaunchArgBuilder[_T]]], type[LaunchArgBuilder[_T]]] | None:
+        """A type safe way to add a mapping of settings builder to launcher to
+        handle the settings at launch time.
+        """
+
+        def register(
+            args_: type[LaunchArgBuilder[_T]], /
+        ) -> type[LaunchArgBuilder[_T]]:
+            if args_ in self._dispatch_registry and not allow_overwrite:
+                launcher_type = self._dispatch_registry[args_]
+                raise TypeError(
+                    f"{args_.__name__} has already been registered to be "
+                    f"launched with {launcher_type}"
+                )
+            self._dispatch_registry[args_] = to_launcher
+            return args_
+
+        if args is not None:
+            register(args)
+            return None
+        return register
+
+    def get_launcher_for(
+        self, args: LaunchArgBuilder[_T] | type[LaunchArgBuilder[_T]], /
+    ) -> type[LauncherLike[_T]]:
+        """Find a type of launcher that is registered as being able to launch
+        the output of the provided builder
+        """
+        if not isinstance(args, type):
+            args = type(args)
+        launcher_type = self._dispatch_registry.get(args, None)
+        if launcher_type is None:
+            raise TypeError(
+                f"{type(self).__name__} {self} has no launcher type to "
+                f"dispatch to for argument builder of type {args}"
+            )
+        # Note the sleight-of-hand here: we are secretly casting a type of
+        # `LauncherLike[Any]` to `LauncherLike[_T]`. This is safe to do if all
+        # entries in the mapping were added using a type safe method (e.g.
+        # `Dispatcher.dispatch`), but if a user were to supply a custom
+        # dispatch registry or otherwise modify the registry THIS IS NOT
+        # NECESSARILY 100% TYPE SAFE!!
+        return launcher_type
+
+
+default_dispatcher: t.Final = Dispatcher()
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
+# -----------------------------------------------------------------------------
+
+
+def create_job_id() -> LaunchedJobID:
+    return LaunchedJobID(str(uuid.uuid4()))
+
+
+class LauncherLike(t.Protocol[_T_contra]):
+    def start(self, launchable: _T_contra) -> LaunchedJobID: ...
+    @classmethod
```

Comment:
oh is this LauncherLike a holder for future launchers?

Reply:
Even better! It's a `Protocol` to allow us to use [structral typing](https://en.wikipedia.org/wiki/Structural_type_system) (often times refered to as [duck typing](https://docs.python.org/3/glossary.html#term-duck-typing) in python land)

Basically this a type that when used as a type hint ostensibly says "I don't care what you give me, as long as it has these methods!"

In that rearguard it is a lot like an `ABC`, but `ABC`s require strict inheritance, which I find limits polymorphism. For e.g.
```py
from abc import ABC, abstractmethod

class MyNumberGetterInterface(ABC):
    @abstractmethod
    def get_number(self) -> int: ...

class FourGetter(MyNumberGetterInterface):
    def get_number(self) -> int: return 4

class SixGetter:  # <-- Note this does not implement the `MyNumberGetterInterface`
    def get_number(self) -> int:  # <-- but does have all of the required methods
        return 6

def use_number(number_getter: MyNumberGetterInterface) -> None: ...
use_number(FourGetter())  # Works!
use_number(SixGetter())  # Should work, but raises a type error :(
```

whereas with a protocol
```py
from abc import abstractmethod
from typing import Protocol

class NumberGetterLike(Protocol):
    @abstractmethod
    def get_number(self) -> int: ...

class FourGetter(NumberGetterLike):
    def get_number(self) -> int: return 4

class SixGetter:  # <-- Note this still does not implement the `NumberGetterLike`
    def get_number(self) -> int:  # <-- but does have all of the required methods
        return 6

def use_number(number_getter: NumberGetterLike) -> None: ...
use_number(FourGetter())  # Works!
use_number(SixGetter())  # Also works!!!
```